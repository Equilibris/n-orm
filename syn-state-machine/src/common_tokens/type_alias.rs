use super::*;
use crate::*;

#[derive(Debug)]
pub struct TypeAlias {
    pub id: Ident,
    pub params: Option<GenericParams>,
    pub bounds: Option<TypeParamBounds>,
    pub pre_where_clause: Option<WhereClause>,

    pub ty: Option<Type>,
    pub post_where_clause: Option<WhereClause>,
}
impl MappedParse for TypeAlias {
    type Source = (
        KwType,
        Identifier,
        Option<GenericParams>,
        Option<(Colon, TypeParamBounds)>,
        Option<WhereClause>,
        Option<(Eq, Type, Option<WhereClause>)>,
        Semi,
    );

    type Output = Self;
    type Error = SmErr<Self::Source>;

    fn map(
        src: SmOut<Self::Source>,
    ) -> Result<<Self as MappedParse>::Output, <Self as MappedParse>::Error> {
        let (ty, post_where_clause) = match src.5 {
            Some((_, a, b)) => (Some(a), b),
            None => (None, None),
        };

        Ok(Self {
            id: src.1,
            params: src.2,
            bounds: src.3.map(|v| v.1),
            pre_where_clause: src.4,
            ty,
            post_where_clause,
        })
    }

    fn map_err(src: SmErr<Self::Source>) -> <Self as MappedParse>::Error {
        src
    }
}

#[cfg(test)]
mod tests {
    use super::TypeAlias;
    use crate::insta_match_test;

    insta_match_test!(it_matches_simple, TypeAlias: type Point = (u8, u8););
    insta_match_test!(it_matches_complex, TypeAlias: type Point<T> where T: std::ops::Add = (T, T););
}
