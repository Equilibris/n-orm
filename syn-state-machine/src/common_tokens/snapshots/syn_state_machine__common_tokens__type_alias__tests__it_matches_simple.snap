---
source: syn-state-machine/src/common_tokens/type_alias.rs
expression: "crate::parse::<TypeAlias>(::quote::quote! { type Point = (u8, u8) ; })"
---
Ok(
    (
        TypeAlias {
            id: Ident(
                Point,
            ),
            params: None,
            bounds: None,
            pre_where_clause: None,
            ty: Some(
                NoBounds(
                    Tuple(
                        TupleType(
                            [
                                NoBounds(
                                    TypePath(
                                        TypePath {
                                            leading: false,
                                            segments: [
                                                Simple {
                                                    id: Id(
                                                        Ident(
                                                            u8,
                                                        ),
                                                    ),
                                                },
                                            ],
                                        },
                                    ),
                                ),
                                NoBounds(
                                    TypePath(
                                        TypePath {
                                            leading: false,
                                            segments: [
                                                Simple {
                                                    id: Id(
                                                        Ident(
                                                            u8,
                                                        ),
                                                    ),
                                                },
                                            ],
                                        },
                                    ),
                                ),
                            ],
                        ),
                    ),
                ),
            ),
            post_where_clause: None,
        },
        0,
    ),
)
